HA historical sensor

shtrom opened the pull request:
Fix for #6 (historical sensor), and also #10 (mfa)

shtrom committed:
autopep linting

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Switch kWh usage to STATE_CLASS_TOTAL_INCREASING to avoid reset artefacts

fixes: #6

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add per-tariff summary

Tariffs 31, 41, 61 and 62 are a wild guess, and totally untested.

fixes: #6 some more
Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Use STATE_CLASS_TOTAL with last_reset for everything

Neither STATE_CLASS_TOTAL_INCREASING nor STATE_CLASS_MEASUREMENT work
well for the ca. 9am reset that seems to happen for daily summary data.

The `last_reset` time is set to whenever a value changes. This means
this is liable to a bug where a value will be missed it it is the same
as the previous day.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Make update asynchronous

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Split out AuroraPlus API coordinator

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Sanitise entity_id

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add HistoricalSensor for daily per-tariff statistics #6

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Fix Throttle behaviour

The AuroraApi proxies everything to the session, or swallows the error
transparently.  This is not the behaviour that the Throttle wrapper
expects, as it needs to use a _throttle attribute on the wrapped object.

We add a special case in the proxy juuuuust for this.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Cleanup logging

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Be a bit more defensive on empty data

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
@LeighCurran I'm still letting this soak on my HA instance (see #6), but I think this is no pretty much good to go.

shtrom commented:
Hum, no... I think my Throttle handling has broken something...

shtrom committed:
Update to auroraplus 1.5.0

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Only create entities for available tariffs

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

slothking87 commented:
Im not sure if its my system, or if its the last commit to show only available tariffs - But I seem to have lost my tariff sensors this morning, and no amount of removing and adding the integration seems to be able to get them back.

shtrom commented:
Dammit. Likely my changes, sorry.

I noticed that the API tends to return no data for a while for the previous day once it ticks over to the new day. That includes the tariffs.

When that's the case, the integration should refuse to start, and get retried later by HA.

Do you have something in your logs that says as much?
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
No need to be sorry  - thats the joys of testing new things!

I cant see anything related to the component in the logs anywhere - Did I need to change the logging level somewhere? I tried in the config.yaml file, but that hasnt seemed to have helped - But I might have the code wrong

slothking87 commented:
I rolled back to the previous version, then tried to roll forward once the data for yesterday was available. Same thing happened where it marked all sensors apart from kilowatt  hour, estimated balance and dollar value usage.

shtrom commented:
Ok, the data's in now. It looks like it takes most of the next day before the API has the next-day data.

Try restarting your instance now, it should now have the some data, including tariffs.

I have an idea to make this more resilient. That'll require adding a feature to the AuroraPlus Python module. I'll have a go at this later today.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom committed:
Look backwards in time for valid data when updating

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
[REVERTME] Point to AuroraPLus fork

Waiting for LeighCurran/AuroraPlus#1 to be merged.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
@slothking87 Don't try the latest commits. They don't work just yet 😉

shtrom commented:
@slothking87 try this one. There's still a bug in that, if you're on a ToU tariff (93), and you start HA on a weekend, it won't create the PEAK sensor. 

Job for later q:

shtrom committed:
Get Tariffs from monthly summary

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
Oh well... After midnight counts as “tomorrow”, right? ToU tariffs should be there even when starting on weekends now.

slothking87 commented:
Hmmm - Initially it shows all my sensors as unavailable again.. Ill try and reset it a bit later once yesterdays data come in to see if that helps. Just to confirm to update Im just downloading the latest files from your fork and replacing the ones already in HA? Just to make sure that its not something that I am doing that is breaking it

![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/44c721a9-3436-4b77-b960-d96f7a913d3f)

shtrom commented:
Yeah, copying all the `custom_components/auroraplus` files to the same path in your `config` directory should do the trick.

It's strange that _nothing_ is reported now (I haven't touched the logic for the daily summaries, so they should still be happy). I suspect you should see errors in your logs.

One thing to note: this is now temporarily using a fork of LeighCurran/AuroraPlus. I updated the `manifest.json`, but I don't know how HA handles this sort of updates. I use a docker container for my tests, so I can trash the whole thing (I did notice discrepancies like using old dep code if I wasn't doing it), but if you're on a more permanent setup, you might have to find where HA install those deps, and how to update them.

My money is on `getoken()` not getting the arguments it expects (which would indicate an old version of either upstream AuroraPlus or my fork branch (where I now fixed that issue).

slothking87 commented:
Ok, after removing the component, removing the config, rebooting, removing any orphaned sensors and statistics, rebooting, then adding the files back in, rebooting, adding the config back in, rebooting Now I have three sensors. 

![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/5c4135f1-54d2-43a0-ad97-e47a931dc77c)

shtrom commented:
No hourly sensors... There should be something in the logs talking about the tariffs it discovered (likely: none).
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom committed:
Add per-tariff daily cost sensors

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Try to improve startup reliability and performance

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
> The logs show these (If these are the correct logs)
> 
> 2023-08-14 21:56:30.406 DEBUG (SyncWorker_5) [custom_components.auroraplus.sensor] Error: None 2023-08-14 21:56:30.407 DEBUG (SyncWorker_5) [custom_components.auroraplus.sensor] Adding sensor: Estimated Balance 2023-08-14 21:56:30.451 DEBUG (SyncWorker_5) [custom_components.auroraplus.sensor] Adding sensor: Dollar Value Usage 2023-08-14 21:56:31.319 DEBUG (SyncWorker_5) [custom_components.auroraplus.sensor] Adding sensor: Kilowatt Hour Usage 2023-08-14 21:56:31.715 DEBUG (SyncWorker_3) [custom_components.auroraplus.sensor] Updating sensor: Estimated Balance 2023-08-14 21:56:31.962 DEBUG (SyncWorker_9) [custom_components.auroraplus.sensor] Updating sensor: Dollar Value Usage 2023-08-14 21:56:35.492 DEBUG (SyncWorker_0) [custom_components.auroraplus.sensor] Updating sensor: Kilowatt Hour Usage
> 
> EDIT: When I changed the password to be deliberately incorrect - it did give me an error in the first line so it is authenticating correctly by the looks

shtrom committed:
Saner log verbosity

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Don't round values in stats

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

slothking87 commented:
I loaded the files again, and the logs look a bit more normal now. Its throwing lots of messages about data not being available which would be expected for this hour of the morning I guess. 
There is also this error that started to come up (in bold below) - Not sure if it was anything to worry about

2023-08-15 08:20:42.369 DEBUG (MainThread) [custom_components.auroraplus.sensor] AuroraApi ready with <auroraplus.api object at 0x7f822eaf90>
**2023-08-15 08:20:42.372 WARNING (SyncWorker_12) [custom_components.auroraplus.sensor] Error updating data: api.gettoken() missing 2 required positional arguments: 'username' and 'password'**
2023-08-15 08:20:42.443 DEBUG (MainThread) [custom_components.auroraplus.sensor] Estimated Balance created
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Dollar Value Usage created
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Kilowatt Hour Usage created
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Kilowatt Hour Usage Tariff T41 created (historical)
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Kilowatt Hour Usage Tariff T31 created (historical)
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Dollar Value Usage Tariff T41 created (historical)
2023-08-15 08:20:42.444 DEBUG (MainThread) [custom_components.auroraplus.sensor] Dollar Value Usage Tariff T31 created (historical)
2023-08-15 08:20:42.444 INFO (MainThread) [custom_components.auroraplus.sensor] Aurora+ platform ready with tariffs ['T41', 'T31']
2023-08-15 08:20:42.483 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for EstimatedBalance
2023-08-15 08:20:42.483 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for EstimatedBalance not yet available
2023-08-15 08:20:42.483 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for DollarValueUsage
2023-08-15 08:20:42.483 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for DollarValueUsage not yet available
2023-08-15 08:20:42.484 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for DollarValueUsage
2023-08-15 08:20:42.484 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for DollarValueUsage not yet available
2023-08-15 08:20:42.484 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for KilowattHourUsage
2023-08-15 08:20:42.484 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for KilowattHourUsage not yet available
2023-08-15 08:20:42.485 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for day
2023-08-15 08:20:42.485 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for day not yet available
2023-08-15 08:20:42.485 WARNING (MainThread) [custom_components.auroraplus.sensor] Kilowatt Hour Usage Tariff T41: no metered records, can't obtain hourly data
2023-08-15 08:20:42.510 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for day
2023-08-15 08:20:42.511 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for day not yet available
2023-08-15 08:20:42.511 WARNING (MainThread) [custom_components.auroraplus.sensor] Kilowatt Hour Usage Tariff T31: no metered records, can't obtain hourly data
2023-08-15 08:20:42.513 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for day
2023-08-15 08:20:42.513 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for day not yet available
2023-08-15 08:20:42.513 WARNING (MainThread) [custom_components.auroraplus.sensor] Dollar Value Usage Tariff T41: no metered records, can't obtain hourly data
2023-08-15 08:20:42.516 DEBUG (MainThread) [custom_components.auroraplus.sensor] Accessing data for day
2023-08-15 08:20:42.516 DEBUG (MainThread) [custom_components.auroraplus.sensor] Data for day not yet available
2023-08-15 08:20:42.516 WARNING (MainThread) [custom_components.auroraplus.sensor] Dollar Value Usage Tariff T31: no metered records, can't obtain hourly data

shtrom commented:
Cool, so now you are on my branch of the integration, but the gettoken error is a tell-tale sign that you still have the upstream AproraPlus Python lib, rather than my fork.

I have the same issue on my actual server. I need to work out how to make HA install a new version, after installing the original one from HACS.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom commented:
HA! This might work https://developers.home-assistant.io/docs/creating_integration_manifest/#custom-requirements-during-development--testing

```
pip install auroraplus@git+https://github.com/shtrom/AuroraPlus@more-than-1-day --target ~/.homeassistant/deps
hass --skip-pip-packages pychromecast
```

(I'm not certain what ~/.homeassistant/deps should actually be in your case)

slothking87 commented:
That sounds plausible. I'll give it a go tomorrow morning. This annoying thing called work keeps getting in the way of all the fun things

shtrom commented:
Ok, my error is different (I think my machine was overloaded due to weekly backups), but the stacktrace helps:
```Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 320, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/auroraplus/sensor.py", line 98, in async_setup_platform
    AuroraPlus = await hass.async_add_executor_job(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/auroraplus/sensor.py", line 96, in aurora_init
    session.getmonth()
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 89, in getmonth
    self.month = self.request("month", index)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 63, in request
    + self.serviceAgreementID
      ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'api' object has no attribute 'serviceAgreementID'
```

I have checked that path, and I can see that my fork is properly installed.

If you check your gettoken error, you should be able to click on the message, and see a stacktrace that lists the path 
![image](https://github.com/LeighCurran/AuroraPlusHA/assets/160280/0a1fb3eb-dfdb-4ca7-bb53-1859eddea614)

That should give you an indication of what path to use with the pip install, in my case, that would have been `/usr/local/lib/python3.11/site-packages/`. That said, I didn't have to do it, as the correct lib was installed after a restart, so maybe you can just get away with removing the `auroraplus` directory and restart.

shtrom committed:
Use absolute values on hourly measurements

This is useful for the T140 feed-in, which is negative, when HA expect a
positive return value.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

slothking87 commented:
Hmmm - I seem to have broken mine completely now. It knows that the files are there in custom_ components because I get the warning in the logs, but it doesnt attempt to get any data which kind of points back to square one on the package problem. I wasnt able to find a way to locate that directory to remove the file (I run everything on a RPI running HAOS core). Ill have another attempt tonight hopefully!

slothking87 commented:
Nope. No luck again. I thought maybe if I removed the integration via hacs that it might remove everything and then reinstall manually - no luck there either. Need to do some more reading

shtrom commented:
It seems to work fine on my real instance now.

I think there are two things you are running into:
* Obviously, HACS reinstalls the upstream version, so you need to copy the files from the branch again
* HA seems to install Python deps on start-up if missing. After installing the integration, I think you need to stop your server, find where the `auroraplus` Python lib (not this integration) is, likely in `config/deps`, or in `.../lib/python.../site-packages` somewhere.

I think if you
1. remove everything,
2. restart,
3. install the upstream integration with HACS,
4. stop the server
5. overwrite the files (mainly `sensor.py` and `manifest.json`)
6. looks for the auroraplus python lib, and remove it
7. restart the server

At this point, it should try to setup the new (patched) integration, realise that a dep is missing, and install it based on the manifest, which should point to my fork.

slothking87 commented:
Ok, new error this time. Unsure if it is related to the same thing. Im still getting the username and password error so I guess that indicates that I have the wrong thing somewhere still.


2023-08-17 18:31:30.696 DEBUG (MainThread) [custom_components.auroraplus.sensor] AuroraApi ready with <auroraplus.api object at 0x7f6b5755d0>
2023-08-17 18:31:30.698 WARNING (SyncWorker_9) [custom_components.auroraplus.sensor] Error updating data: api.gettoken() missing 2 required positional arguments: 'username' and 'password'
.........
2023-08-17 18:31:31.017 ERROR (MainThread) [homeassistant.components.sensor] auroraplus: Error on device update!
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 576, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 938, in async_device_update
    await self.async_update()
  File "/config/custom_components/auroraplus/sensor.py", line 265, in async_update
    float(self._api.EstimatedBalance), self._rounding)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: float() argument must be a string or a real number, not 'dict'

shtrom commented:
Hum, so you still have the gettoken error, but also a dict-instead-of-number that happens in one of the error cases when data is not available (which it isn't because of the gettoken).

The former should no longer happen if on my fork. Are you sure you have the latest version of the manifest.json, and it's in the right place?

As a workaround, you could continue the hacking and manually replace the files froc he AuroraPlus Python lib with my fork, too.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
Getting somewhere slowly. Manually updated the sensor.py file with the latest one and get the same error you got before. Now just to find how to access      /usr/local/lib/python3.11/site-packages/auroraplus

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 362, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/auroraplus/sensor.py", line 98, in async_setup_platform
    AuroraPlus = await hass.async_add_executor_job(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/auroraplus/sensor.py", line 96, in aurora_init
    session.getmonth()
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 72, in getmonth
    self.month = self.request("month")
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 52, in request
    request = self.session.get(self.url + '/usage/' + timespan +'?serviceAgreementID=' + self.serviceAgreementID + '&customerId=' + self.customerId + '&index=-1', headers={'Authorization': self.token})
                                                                                         ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'api' object has no attribute 'serviceAgreementID'

shtrom commented:
Did you do the same with the manifest.json?
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom commented:
On another note, Aurora seems to be stuck at the 14th. I don't have any data since then, even on their web dashboard.

The latest version of the integration should at least be able to get back to the latest day with some data.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
> Did you do the same with the manifest.json?
> -- 
I do - it references the new link.   There is a file I found with the haacs data in there. I might try removing that in case it keeps downloading the original files each reboot. I manually tried to change it to your repository and the new fork but that kept reverting back.

shtrom commented:
What's the path and contents of that file?

slothking87 commented:
\config\.storage\hacs.repositories

 "647237192": {
      "repository_manifest": {
        "name": "Aurora+",
        "render_readme": true,
        "country": [
          "AU"
        ]
      },
      "full_name": "shtrom/AuroraPlusHA",
      "authors": [
        "@LeighCurran"
      ],
      "category": "integration",
      "description": "Home Assistant integration for Aurora+  ",
      "domain": "auroraplus",
      "etag_repository": "W/\"7402aaa2037ecd065622bffdd49b92024fa389f2e84c40c1f2dd64a7802d5c45\"",
      "last_updated": "2023-08-15T12:51:56Z",
      "default_branch": "main",
      "installed_commit": "fe4b60f",
      "installed": true,
      "last_commit": "fe4b60f",
      "manifest_name": "Aurora+",
      "last_fetched": 1692235359.467871
    },

slothking87 commented:
Ok, I started again and now I get a new error. Just for something different

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 576, in _async_add_entity
    await entity.async_device_update(warning=False)
  File "/usr/src/homeassistant/homeassistant/helpers/entity.py", line 938, in async_device_update
    await self.async_update()
  File "/config/custom_components/auroraplus/sensor.py", line 265, in async_update
    float(self._api.EstimatedBalance), self._rounding)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: float() argument must be a string or a real number, not 'dict'

Still getting the missing arguments message also, so that could be the problem still. 

![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/cde98885-0fb6-48f4-91dd-be64c57aebc1)


![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/5d65be17-881c-4315-9c4b-fe752e14d2fb)

shtrom commented:
Yeah, I think that's the same cause.

Have you found where HA installs the Python deps? I think it might be worth trying to upgrade your install of AuroraPlus with my branch (see previous message about using `pip`). I doubt HA will try to replace it, which I suspect is also what's preventing the update.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
Yeah - I've been trying to find a way to use that. I run home assistant supervised on a RPI which seems to make it  more challenging

shtrom commented:
Do you have console/ssh access to it?
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
I can SSH into the core. But the commands dont seem to work in that section. There doenst seem to be much info out there on how all that works.

slothking87 commented:
Ok moving somewhere. I found an add on that lets me install custom packages  (https://github.com/hassio-addons/addon-appdaemon). 

Down to the single error now
![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/b9986b8b-f1ee-453f-8c9e-5b3712311e94)


The logs from the install below if that means anything much:
Collecting https://github.com/shtrom/AuroraPlus/archive/refs/heads/more-than-1-day.zip
  Downloading https://github.com/shtrom/AuroraPlus/archive/refs/heads/more-than-1-day.zip (6.7 kB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'done'
Requirement already satisfied: requests in /usr/lib/python3.11/site-packages (from auroraplus==1.1.6) (2.28.2)
Requirement already satisfied: charset-normalizer<4,>=2 in /usr/lib/python3.11/site-packages (from requests->auroraplus==1.1.6) (3.2.0)
Requirement already satisfied: idna<4,>=2.5 in /usr/lib/python3.11/site-packages (from requests->auroraplus==1.1.6) (3.4)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in /usr/lib/python3.11/site-packages (from requests->auroraplus==1.1.6) (1.26.16)
Requirement already satisfied: certifi>=2017.4.17 in /usr/lib/python3.11/site-packages (from requests->auroraplus==1.1.6) (2023.7.22)
Building wheels for collected packages: auroraplus
  Building wheel for auroraplus (setup.py): started
  Building wheel for auroraplus (setup.py): finished with status 'done'
  Created wheel for auroraplus: filename=auroraplus-1.1.6-py3-none-any.whl size=4376 sha256=8e78b487be1c07af179ab15202aeb9440073369bb880c3de645da421155d4ada
  Stored in directory: /tmp/pip-ephem-wheel-cache-zdr8traz/wheels/c3/4c/f9/fbb5d2c86727f1c5c19439e058ea9389c1efee140b57b6c3a3
Successfully built auroraplus
Installing collected packages: auroraplus
Successfully installed auroraplus-1.1.6

slothking87 commented:
> [slothking87](/slothking87)

Nevermind - Read the logs again and the authentication error was there - Just hidden a bit deeper

slothking87 commented:
SUCCESS!! If anyone else as noob as me is reading this - The  way to run the pip command is to install the Advanced SSH & Web Terminal add on. The other SSH add on I was using didnt allow the correct access to run the commands. 

After running the command and rebooting it seems to have grabbed the backdated data fairly instantly.

slothking87 commented:
> Woohoo! You would just have gotten one day worth of data, right? -- Olivier Mehani ***@***.***> Sent from my mobile, please excuse my brevity.

It seems to have pulled everything back to the 19th when I check the sensors and the logs. 

It does show this under the dev tools/statistics section. Previously if I deleted them and rebooted it seemed to work ok, but not this time. Are you seeing this in yours also? I might just be impatient as it has only just started to pull down data. 

![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/4106773f-a2a6-4e0b-82fb-93a376bb70f6)


![image](https://github.com/LeighCurran/AuroraPlusHA/assets/54494142/d9dc2ab4-fc6f-49ad-a6b8-cc91ee7cae4d)

shtrom commented:
Ah, you must have been lucky and data for the 20th wasn't available when you first started it, so it went back one more day.

The `Unsupported state class` is, AFAIK, expected, as this is how ha-historical-sensor works: it doesn't have a state class (for immediate data), and instead calculates stats itself and injects them into the recorder.

Judging by the comments in the example, setting a state class would break stats, so I think we'll have to live with this warning. But I think it's harmless.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
That makes sense. So far nothing has imported into the energy dashboard. It has recorded my cakes up to the 13 when I started trying to make changes, then nothing. I guess I'll see if it imports tomorrow morning with the new data. At least I can see it in the sensor information

shtrom commented:
They should show up as soon as recorded.

Make sure you are using the right entities. You might have to remove them from the Energy config, delete the unavailable one, again, and re-do the config.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
That did the trick. I had to remove the config, reboot, remove the old statistics and sensors, reboot, put the config back in and reboot once more and that gave home the correct things to put in the energy dashboard. It cleared any history I had in there - but it at least works now!

shtrom commented:
Cool. I think you're good to go now!

I've had it running for about a week with no new changes required, so I think it's pretty solid now. The only issue seems to be data in Aurora being slow to show up (up to multiple days).

Let's see how you fare. Barring install issues that you had due to running forked code, the rest should be smooth now.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
Yep. All looking good so far - Thanks heaps for your help (and patience). Out of interest how far back should it be grabbing historical data?

shtrom commented:
tl;dr: it only looks as far as the most recent day with data.

The Aurora+ API lags a bit before it has data, so it tells you whether data is available or not for each day.

If data is not available, the AuroraPlusHA historical sensors will try the day before, all the way to the latest day that has some data available. If the AuroraPlus API makes multiple days available at once (or within an hour, which is the refresh period), the integration will only look at the latest, and not go backwards any further.

The HistoricalState then only insert data that is newer than the last recorded data, and ignores older data, even if it was previously missing.

There might be a way to play further with recorder to 1) find gaps in statistics, 2) query those from the Aurora+ API and 3) backfill the data in he recorder. However, I don't think it's natively supported by the HistoricalState, so would be tricky to implement. 

I concluded I could live with some occasional gaps (;
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom committed:
Update AuroraPlus branch for MFA

Usage:

1. obtain an `access_token` following instructions at [0]
2. update the secrets to use an empty username (`''`), and the
   `access_token` as the password

[0] https://github.com/shtrom/AuroraPlus/tree/oauth-mfa-token\#obtain-a-token

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
@slothking87 I just pushed an interim fix to support MFA https://github.com/shtrom/AuroraPlus/tree/oauth-mfa-token to this branch.

Instructions are in this commit https://github.com/LeighCurran/AuroraPlusHA/pull/9/commits/7e277e5db346cba05effbb02f2d0cd5151773c8c

You might have to redo the dance to trash your previous state and pull a new version of the dependency library

slothking87 commented:
I feel like Im missing something very obvious - But I cant get it to work. 

When loading the URL mentioned in a web browser, I do get a generic error page, but with this message which makes me think Im missing something somewhere

"Sorry, but we're having trouble signing you in.
We track these errors automatically, but if the problem persists feel free to contact us. In the meantime, please try again.
Correlation ID: 2ba71a8f-221c-4b05-8d0d-1e1d6ade01c9
Timestamp: 2023-11-29 06:14:55Z
AADB2C99059: The supplied request must present a code_challenge"

shtrom commented:
Yeah, the url should have been generated with a code challenge...

Can you show me the code you used, and the url (this one is safe to share as it's protected by the code_challenge so it's locked to your code, but feel free to edit parts out if you want)
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
Excuse the noob questions - So the process is to run the commands as they are listed and that opens the correct link? 
The code I got was identical to yours so guessing I missed something or something didnt want to work.

shtrom commented:
Ok, so this URL should be pasted back into the `api.redirect_uri()`, and you'll get an access_token.

Then, yes, update your HA secrets to have `username=''`, and `password=<access_token>`, and restart HA. If you're in luck, you'll start getting data again.

shtrom commented:
PS: be careful about posting `redirect_uri`s (I was talking about posting the first URL that is returned by `oauth_authorize`), as they do contain secrets. 

In this case, the secret is protected by the `code_challenge` that only your client `api` instance has, but best be prudent.

shtrom commented:
> If you're in luck, you'll start getting data again.

I am! 

![image](https://github.com/LeighCurran/AuroraPlusHA/assets/160280/d87034d7-3561-4420-b7e1-d94a0e0bcec3)

slothking87 commented:
so close! I got the extra information, but getting the usual authentication error inside home assistant. How much of that final string is whats needed for the access token? 

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 359, in _async_setup_platform
    await asyncio.shield(task)
  File "/config/custom_components/auroraplus/sensor.py", line 98, in async_setup_platform
    AuroraPlus = await hass.async_add_executor_job(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/auroraplus/sensor.py", line 96, in aurora_init
    session.getmonth()
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 89, in getmonth
    self.month = self.request("month", index)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/auroraplus/__init__.py", line 63, in request
    + self.serviceAgreementID
      ^^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'api' object has no attribute 'serviceAgreementID'

shtrom commented:
Ok, same thing here on my non-test HA deployment. Hold tight.

shtrom commented:
You need to redo what you did there https://github.com/LeighCurran/AuroraPlusHA/pull/9#issuecomment-1685250663 to get the new version of the auroraplus lib.

shtrom commented:
Ok, that didn't work. But I realised existing instances do need an update in the `config/custom_components/auroraplus/manifest.json` to point to the new branch: https://github.com/shtrom/AuroraPlusHA/commit/7e277e5db346cba05effbb02f2d0cd5151773c8c, then a restart.

I also had to delete and remove and reinstall the previous install of the `auroraplus`:

```
rm -rf /usr/local/lib/python3.11/site-packages/auroraplus*
pip install --upgrade --no-cache-dir   git+https://github.com/shtrom/AuroraPlus@oauth-mfa-token 
```

You can check with

```
cat /usr/local/lib/python3.11/site-packages/auroraplus/__init__.py
```

that it corresponds to the version at https://github.com/shtrom/AuroraPlus/blob/oauth-mfa-token/auroraplus/__init__.py

slothking87 commented:
Hmmm, Im still getting the same error as before with the no service agreement ID. Ill have a better play over the weekend - Im probably doing something wrong somewhere

shtrom commented:
As usual, Aurora had a wobble just when I was testing, but it came good today, so it works on mine.

If you still have the `serviceAgreementID`, it's likely you're not on the updated branch yet. I'm not sure how you managed to update it last time, but if you can ssh into the system that runs it, check the `__init__.py` and compare it to the branch. If it still is different, this means that the lib is not updated.

Worse case scenario, just delete the `auroraplus` directory (the Python lib, not the HA integration), and it should get reinstalled, hopefully to the correct version, on restart.

slothking87 commented:
Bingo - That worked!!!!!! 

IF anyone uses supervised HA on a RPI like I do, You can run the pip commands via an add on (https://github.com/home-assistant/addons-development).  Just add the command (pip install git+https://github.com/shtrom/AuroraPlus@oauth-mfa-token in this case) under the pypi option, save, then start the add on and it does the rest for you. keep an eye on the logs tab  to see when its done.

shtrom commented:
Awesome!

I still need to work out how to make HA fetch and record the token, rather than having to put it in manually, but that's a good functional workaround in the meantime.

shtrom commented:
Ok, it looks like the `access_token` has a three-week expiry. Data stopped coming on the 20th, I just had to put a new one in my `secrets.yaml` and restart. I got data for yesterday straight away, as expected. Let's see if that expires around the 17th.

If so, that'll make the need for a way to refresh the token more pressing... If that's indeed something that the Aurora+ Web UI supports. Otherwise, we'll have to at least streamline the process of obtaining a new token.

shtrom committed:
Bump homeassistant-historical-sensor to 2.0.0rc5

Fixes

    Traceback (most recent call last):
    File /usr/src/homeassistant/homeassistant/loader.py, line 842, in get_platform
    cache[full_name] = self._import_platform(platform_name)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File /usr/src/homeassistant/homeassistant/loader.py, line 859, in _import_platform
    return importlib.import_module(f{self.pkg_path}.{platform_name})
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File /usr/local/lib/python3.11/importlib/init.py, line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File , line 1204, in _gcd_import
    File , line 1176, in _find_and_load
    File , line 1147, in _find_and_load_unlocked
    File , line 690, in _load_unlocked
    File , line 940, in exec_module
    File , line 241, in _call_with_frames_removed
    File /config/custom_components/auroraplus/sensor.py, line 39, in
    from homeassistant_historical_sensor import (
    File /usr/local/lib/python3.11/site-packages/homeassistant_historical_sensor/init.py, line 39, in
    raise SystemError(msg)
    SystemError: Running HomeAssistant 2024.1.2, Minimum required version >=2023.6.0

See [0], [1]

[0] ldotlopez/ha-historical-sensor#12
[1] https://github.com/LeighCurran/AuroraPlus/pull/3\#issuecomment-1884213483

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
@slothking87 re: https://github.com/LeighCurran/AuroraPlus/pull/3#issuecomment-1884213483

I just updated the ha-historical-sensor as per https://github.com/ldotlopez/ha-historical-sensor/issues/12, and pushed it.

This should fix your issue on 2024.1.2, but you'll have to do the usual dance of trashing the old python dep to and make sure the new one is installed.

shtrom commented:
> Ok, it looks like the `access_token` has a three-week expiry. Data stopped coming on the 20th, I just had to put a new one in my `secrets.yaml` and restart. I got data for yesterday straight away, as expected. Let's see if that expires around the 17th.

It lasted until the 25th! Sound like a... 29 day lifetime? I'm renewing it today, so I guess we'll see on the 26th of Feb?

shtrom commented:
(psst, have a try at #11 for a UI-based configuration of the access token).
